### PR TITLE
Validate dates in `.datetime()`

### LINKED ISSUE #2357: .datetime() validation
Hello, the current implementation for z.string().datetime() passes validation with incorrect values.

```javascript
const valid_date = z.string().datetime().parse('1234-12-12T12:12:12.123Z'); // valid.
const date_with_max_year = z.string().datetime().parse('9999-12-12T12:12:12.123Z'); // valid.

// Invalid time.
const date_with_invalid_seconds = z.string().datetime().parse('1234-12-12T12:12:99.123Z'); // valid, should throw error -> seconds can't be > 60
const date_with_invalid_minutes = z.string().datetime().parse('1234-12-12T12:99:12.123Z'); // valid, should throw error -> minutes can't be >= 60
const date_with_invalid_hours = z.string().datetime().parse('1234-12-12T99:12:12.123Z'); // valid. should throw error -> hours can't be > 24

// Invalid date.
const date_with_invalid_day = z.string().datetime().parse('1234-12-99T12:12:12.123Z'); // valid, should throw error -> day can't be > 31
const date_with_invalid_month = z.string().datetime().parse('1234-99-12T12:12:12.123Z'); // valid, should throw error -> month can't be > 12

console.log(new Date(valid_date)); // 1234-12-12T12:12:12.123Z
console.log(new Date(date_with_max_year)); // 1234-12-12T12:12:12.123Z

console.log(new Date(date_with_invalid_day)); // Invalid date.
console.log(new Date(date_with_invalid_month)); // Invalid date.

console.log(new Date(date_with_invalid_seconds)); // Invalid date.
console.log(new Date(date_with_invalid_minutes)); // Invalid date.
console.log(new Date(date_with_invalid_hours)); // Invalid date.
```

The same issue might be in ZodString.date() & ZodString.time() implementation (https://github.com/colinhacks/zod/pull/1766)

### PR DESCRIPTION
**Pull request by samchungy**
Resolves #2357

I think this change makes sense. If your JS system using Zod is going to process a datetime string, it will likely use `new Date()` to process it at some point in the system, just maybe not right now. So we should validate that `new Date()` works on it to get rid of things like invalid months, invalid days, invalid seconds etc.

However, this change is technically a small breaking change as https://github.com/colinhacks/zod/pull/1797#issue-1516383731 introduced a date time format which `new Date()` does not accept which seems like a strange addition. I would suggest that users using that format should use a custom regex to validate instead as I will assume the majority of users would not be using that format.

### CODE REVIEW

File: deno/lib/__tests__/string.test.ts [hunk 1]
```diff
@@ -435,7 +440,6 @@ test("datetime parsing", () => {
   datetimeOffset.parse("2020-10-14T17:42:29+00:00");
   datetimeOffset.parse("2020-10-14T17:42:29+03:15");
   datetimeOffset.parse("2020-10-14T17:42:29+0315");
-  datetimeOffset.parse("2020-10-14T17:42:29+03");
```
Comment (samchungy) [inline deno/lib/__tests__/string.test.ts:438]: These aren't parseable by new Date().

**Comment by peterjdrb** [comment]
I've been discussing using Zod at work and whilst one of my collegues was doing a proof of concept, he noticed is was accepting dates such as the 99th of June. Looked at the code as realised it was just doin some regex. Whilst this was good start, it doesn't cover the scenario mentioned above.

@samchungy - your work is still an improvement but there are some cases that you are missing. For example it will accept a the 29th of Febuary 2021 even though this is not a valid year. Also I don't think that the day will validate for the correct months. I.e. if you put the 31st September then the validaiton will fail.

I have already added some tests and written some code on your branch @samchungy. Are you happy for me to commit and push up for you to review?

**Comment by samchungy** [comment]
> I've been discussing using Zod at work and whilst one of my collegues was doing a proof of concept, he noticed is was accepting dates such as the 99th of June. Looked at the code and realised it was just doing some regex. Whilst this was good start as it checks the format, it doesn't cover a number of scenarios including the one mentioned above.
> 
> @samchungy - your work is still an improvement but there are some cases that you are missing. For example it will accept a the 29th of Febuary 2021 even though this is not a valid year. Also I don't think that the day will validate for the correct months. I.e. if you put the 31st September then the validaiton will not fail when it should
> 
> I have already added some tests and written some code on your branch @samchungy. I will push these changes up for you to review

Hey mate, I went a slightly different approach but I stole some of your test cases

**Comment by colinhacks** [comment]
Okay, I've rolled back that change you mentioned @samchungy. Good catch. All successful results should definitely be parsable by `new Date()`.

As for this particular PR, it's much slower in Node.js than the regex-based approach. I added a benchmark in https://github.com/colinhacks/zod/commit/35f0a3804fdad33c8c5cb659aaf9a6ce922038a2 that can be run with `yarn benchmark --datetime`.


```sh
$ yarn benchmark --datetime
yarn run v1.22.19
warning ../../../package.json: No license field
 $ tsx src/benchmarks/index.ts --datetime
datetime: new Date() x 7,195,365 ops/sec ±1.23% (96 runs sampled)
datetime: regex validation x 29,761,251 ops/sec ±0.42% (96 runs sampled)
datetime: simple regex with validation x 6,113,522 ops/sec ±0.66% (97 runs sampled)
```

Given that I'm inclined to stick with regex. But Sam if you see any glaring issues with that benchmark let me know 👍

**Comment by samchungy** [comment]
> Okay, I've rolled back that PR you mentioned @samchungy. Good catch. All successful results should definitely be parsable by `new Date()`.
> 
> As for this particular PR, it's much slower in Node.js than the regex-based approach. I added a benchmark in [35f0a38](https://github.com/colinhacks/zod/commit/35f0a3804fdad33c8c5cb659aaf9a6ce922038a2) that can be run with `yarn benchmark --datetime`.
> 
> ```shell
> $ yarn benchmark --datetime
> yarn run v1.22.19
> warning ../../../package.json: No license field
>  $ tsx src/benchmarks/index.ts --datetime
> datetime: new Date() x 7,195,365 ops/sec ±1.23% (96 runs sampled)
> datetime: regex validation x 29,761,251 ops/sec ±0.42% (96 runs sampled)
> datetime: simple regex with validation x 6,113,522 ops/sec ±0.66% (97 runs sampled)
> ```
> 
> Given that I'm inclined to stick with regex. But Sam if you see any glaring issues with that benchmark let me know 👍

Oooo yeah that is a significant degradation. Let me cook something up and see if I can make it better

**Comment by samchungy** [comment]
Latest master covers all of my new test cases so I'm going to close this.

**Pull request closed without merging**